handlers/handlers.go

```diff
@@ -1,238 +1,238 @@
-					http.StatusTemporaryRedirect)
-			} else {
-				authenticatedHandlerFunc(responseWriter, request, userId)
-			}
-		}
-}
-
-// Authenticated Handlers
-func HandleHomeRequest(
-	responseWriter http.ResponseWriter,
-	request *http.Request,
-	userId models.UserId,
-) {
-	switch request.Method {
-	case http.MethodGet:
-		parsedTemplate, err := template.ParseFiles("templates/home.tmpl")
-		if err != nil {
-			log.Fatal(err)
-		}
-
-		parsedTemplate.Execute(responseWriter, userId)
-	default:
-		respondWithMethodNotAllowed(responseWriter, []string{http.MethodGet})
-	}
-}
-
-// UTIL
-func respondWithMethodNotAllowed(
-	responseWriter http.ResponseWriter,
-	allowedMethods []string,
-) {
-	allowedMethodsString := strings.Join(allowedMethods, ", ")
-	responseWriter.Header().Set("Allow", allowedMethodsString)
-
-	statusCode := http.StatusMethodNotAllowed
-	http.Error(responseWriter, http.StatusText(statusCode), statusCode)
-}
-
-func getRequestBody(request *http.Request) []byte {
-	body, err := ioutil.ReadAll(request.Body)
-	if err != nil {
-		panic(err)
-	}
-
-	if err := request.Body.Close(); err != nil {
-		panic(err)
-	}
-
-	return body
-}
+package handlers 
+ 
+import ( 
+	"encoding/json" 
+	"fmt" 
+	"github.com/atmiguel/cerealnotes/models" 
+	"github.com/atmiguel/cerealnotes/paths" 
+	"github.com/atmiguel/cerealnotes/services/userservice" 
+	"github.com/dgrijalva/jwt-go" 
+	"html/template" 
+	"io/ioutil" 
+	"log" 
+	"net/http" 
+	"strings" 
+	"time" 
+) 
+ 
+const oneWeekInMinutes = 60 * 24 * 7 
+const credentialTimeoutInMinutes = oneWeekInMinutes 
+const cerealNotesCookieName = "CerealNotesToken" 
+ 
+type JwtTokenClaim struct { 
+	models.UserId `json:"userId"` 
+	jwt.StandardClaims 
+} 
+ 
+var tokenSigningKey []byte 
+ 
+func SetTokenSigningKey(key []byte) { 
+	tokenSigningKey = key 
+} 
+ 
+// HANDLERS 
+func HandleLoginOrSignupRequest( 
+	responseWriter http.ResponseWriter, 
+	request *http.Request, 
+) { 
+	switch request.Method { 
+	case http.MethodGet: 
+		// Check to see if user is already logged in, if so redirect 
+		if _, err := getUserIdFromJwtToken(request); err == nil { 
+			http.Redirect( 
+				responseWriter, 
+				request, 
+				paths.HomePath, 
+				http.StatusTemporaryRedirect) 
+			return 
+		} 
+ 
+		parsedTemplate, err := template.ParseFiles("templates/login_or_signup.tmpl") 
+		if err != nil { 
+			log.Fatal(err) 
+		} 
+ 
+		parsedTemplate.Execute(responseWriter, nil) 
+ 
+	default: 
+		respondWithMethodNotAllowed(responseWriter, []string{http.MethodGet}) 
+	} 
+} 
+ 
+func HandleUserRequest( 
+	responseWriter http.ResponseWriter, 
+	request *http.Request, 
+) { 
+	type SignupForm struct { 
+		DisplayName  string `json:"displayName"` 
+		EmailAddress string `json:"emailAddress"` 
+		Password     string `json:"password"` 
+	} 
+ 
+	switch request.Method { 
+	case http.MethodPost: 
+		var signupForm SignupForm 
+		body := getRequestBody(request) 
+ 
+		if err := json.Unmarshal(body, &signupForm); err != nil { 
+			panic(err) 
```

Comment:
These panics are pretty bad
1. panic doesn't just kill the current goroutine, it crashes the whole program. So one user's malformed request could crash your whole server instead of just 400ing to that user. Be very wary of panics. panic should only be used in cases where the only reasonable thing to do is crash the whole service. I've only ever used panic very early in the startup of a service (e.g. failure to load the config file), or in places where something is logically impossible (e.g. in the default of a switch statement that should never occur).

2. You _could_ [defer a recover](https://blog.golang.org/defer-panic-and-recover) at the top of this function. That would work, but it would be non-idomatic. The idiomatic thing to do is to immediately handle the error and return, or bubble the error up to the caller (in this case there is no caller). Obvs if the error handling gets complicated or duplicated a bunch, then factor the error handling out into its own function